drivers/SmartThings/matter-switch/src/init.lua

```diff
@@ -501,9 +583,73 @@ local function occupancy_attr_handler(driver, device, ib, response)
   device:emit_event(ib.data.value == 0x01 and capabilities.motionSensor.motion.active() or capabilities.motionSensor.motion.inactive())
 end
 
+local function power_topology_handler(driver, device, ib, response)
+  -- check get_endpoints for each feature map value.
+  for name, bitmap in pairs(POWER_TOPOLOGY_FEATURE_MAP) do
+    local feature_eps = embedded_cluster_utils.get_endpoints(device, clusters.PowerTopology.ID, {feature_bitmap = bitmap})
+    if #feature_eps > 0 then
+      device:set_field(POWER_TOPOLOGY_TYPE, POWER_TOPOLOGY_FEATURE_MAP[name])
+    end
+  end
+end
+
+local function cum_energy_exported_handler(driver, device, ib, response)
+  device:set_field(TOTAL_EXPORTED_ENERGY, ib.data.elements["Energy"])
+  device:set_field(LATEST_EXPORTED_TIMESTAMP, ib.data.elements["EndTimestamp"])
+  device:emit_event(capabilities.energyMeter.energy({ value = ib.data.elements["Energy"], unit = "Wh" }))
+end
+
+local function per_energy_exported_handler(driver, device, ib, response)
+  local latest_energy_report = device:get_field(TOTAL_EXPORTED_ENERGY)
+  local summed_energy_report = latest_energy_report + ib.data.elements["Energy"]
+  device:set_field(TOTAL_EXPORTED_ENERGY, summed_energy_report)
+  device:set_field(LATEST_EXPORTED_TIMESTAMP, ib.data.elements["EndTimestamp"])
+  device:emit_event(capabilities.energyMeter.energy({ value = ib.data.elements["Energy"], unit = "Wh" }))
+end
+
+local function energy_report_handler_factory(is_cumulative_report)
+  return function(driver, device, ib, response)
+    if not device:get_field(REPORT_POLL_TIMER_IS_SET) then
+      local cumulative_eps = embedded_cluster_utils.get_endpoints(device, clusters.ElectricalEnergyMeasurement.ID,
+        {feature_bitmap = clusters.ElectricalEnergyMeasurement.types.ElectricalEnergyMeasurementFeature.CUMULATIVEENERGY}
+      )
+      if #cumulative_eps > 1 then
+        device:set_field(CUMULATIVE_ENERGY_IS_REPORTED, true)
+      end
+      if device:get_field(SUBSCRIPTIONS_COMPLETE) then -- with this, we know this must have been called after the subscriptions.
+        local reportInterval = ib.data.elements["EndTimestamp"] - ib.data.elements["StartTimestamp"]
```

Comment:
Just a heads up `startTimestamp` is omitted if the energy measurement is cumulative: https://librarian-regionals.smartthingsgdev.com/matter/Matter_1-3/appclusters.html#ref_EnergyMeasurementStartTimestamp

I think we can just use the 15 minute interval, since the device should send an update whenever the value changes since we have a subscription to this attribute, so it wouldn't delay sending the value.

Reply:
I understand that position, though sending the report at the time frame that is more accurate to the device would be better, no? It would also lower cloud costs that were mentioned as a contributing factor to the 15 minute minimum. For example, even a 17 minute timer vs 15 minute timer would bring down sending by 13%.

Reply:
And yes :sweat_smile: I realized the StartTimestamp is omitted while writing these responses to your reviews. As can be seen in a few places, this one is definitely a first draft. However, I don't think it would be all that difficult to take the difference between the first two event emissions and call that the time interval. A little more overhead than the current implementation, but not too bad.

Reply:
Is there any case that the reporting interval for energy usage of the device may vary?
If the first report was made in an hour, and it occurs every 15 minutes from then on, the device sends energy usage every 15 minutes, but SmartThings will update it every hour. This reduces the accuracy of energy measurements.
Since the device may send an energy reporting at constant interval or an irregular interval, driver would be better to report energy at regular intervals.
And if there is no change in the value of previousTotalConsumptionWh, it would be nice to add a process that does not send an event.